### ISSUE TITLE
Lose performance between 0.6.0 and 0.7.1

### ISSUE THREAD

**Issue by mpariente** [PyTorchLightning/pytorch-lightning#1136]
## 🐛 Bug

When I train exactly the same model with pl 0.7.1, I get worse performance compared to pl0.6.0. 
I did a fresh install or [Asteroid]( https://github.com/mpariente/asteroid) with both versions and ran exactly the same script on the same hardware. 
I get significantly worse performance with pl0.7.1. 
Are there some known issues I should be aware of? In the mean time, I'll have to downgrade to 0.6.0

### Environment
#### PL 0.6.0
Collecting environment information...                                                                                                                                           [8/105]
PyTorch version: 1.4.0                                                                                                                                                                 
Is debug build: No                                                                                                                                                                     
CUDA used to build PyTorch: 10.1                                                                                                                                                       
                                                                                                                                                                                       
OS: Debian GNU/Linux 10 (buster)                                                                                                                                                       
GCC version: (Debian 8.3.0-6) 8.3.0                                                                                                                                                    
CMake version: version 3.14.0                                                                                                                                                          
                                                                                                                                                                                       
Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] pytorch-lightning==0.6.0
[pip3] torch==1.4.0
[pip3] torchvision==0.4.2
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      243
[conda] mkl-include               2020.0                      166
[conda] mkl-service               2.3.0            py36he904b0f_0
[conda] mkl_fft                   1.0.14           py36ha843d7b_0
[conda] mkl_random                1.1.0            py36hd6b4f25_0
[conda] torch                     1.3.1                    pypi_0    pypi
[conda] torchvision               0.4.2                    pypi_0    pypi

#### Diff between 0.6.0 and 0.7.1 envs
diff env_0.7 env_0.6
```
19c19
< [pip3] pytorch-lightning==0.7.1
---
> [pip3] pytorch-lightning==0.6.0
```

**Comment by awaelchli** [comment]
Could you be more precise what you mean with performance? 
It could mean
- Your test loss is lower with 0.7.1 compared to 0.6.0
- Training epochs take longer to finish in 0.7.1 vs. 0.6.0
- or other things.

**Comment by mpariente** [comment]
Yeah true, sorry.
Training, validation and testing losses are worse with 0.7.1, that's what I
meant.

**Comment by Borda** [comment]
@mpariente could you pls give us some numbers?

**Comment by mpariente** [comment]
Yes. We minimize negative signal to distortion ratio (SDR), which is widely used in speech separation. I report loss values here, they are negative and lower is better. 
With 0.6.0, we reached -18dB / With 0.7.1, we couldn't reach better than -15dB. People publish for 1dB so this is a huge difference. 

I'm re-running the experiments with exactly the same version of the code, same libraries, same hardware, only difference is lightning. This is a screenshot of the ongoing runs.We can already see a 2dB difference between the runs
![Screenshot from 2020-03-13 14-19-08](https://user-images.githubusercontent.com/18496796/76624983-9c1ce480-6536-11ea-8532-8e8bfc0435c9.png)

**Comment by Borda** [comment]
just crossed some other comment about slower performances https://github.com/PyTorchLightning/pytorch-lightning/issues/525#issuecomment-596963253

**Comment by gwichern** [comment]
I also noticed a similar issue with my code after upgrading to 0.7.1 from 0.6, so I tried running the [MNIST example](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/basic_examples/cpu_template.py), and confirmed performance difference (both of my environments used torch==1.4.0 and torchvision==0.5.0)

<img width="441" alt="Screen Shot 2020-03-13 at 11 35 06 AM" src="https://user-images.githubusercontent.com/152977/76636188-01fd7200-651f-11ea-9379-2cb2fda197db.png">

 Orange curve is version 0.6, pink curve is version 0.7.1

**Comment by awaelchli** [comment]
If it even happens with MNIST, we have to find the bug asap!
I thought there were tests in place to ensure the performance does not change, but it seems this is not the case.

**Comment by awaelchli** [comment]
Just to be sure, @mpariente and @gwichern did you make the training deterministic before running this comparison?

**Comment by williamFalcon** [comment]
there are test cases to make sure performance doesn't change. please rerun using the exact same seed and only change the versions.

**Comment by gwichern** [comment]
In my case yes.  The pytorch-lightning MNIST example sets both the numpy and torch seeds to 2334.  Everything is the same except environments.

**Comment by williamFalcon** [comment]
I'm not sure what you mean here.
You still reach the same accuracy just at diffferent times 

![image](https://user-images.githubusercontent.com/3640001/76645836-29106f80-6530-11ea-9b16-32090fae184e.png)

**Comment by williamFalcon** [comment]
@gwichern can you post a colab here? we can test it there.

**Comment by gwichern** [comment]
Agreed, on that point about the accuracy getting to the same value, but I just ran pl_examples/basic_examples/cpu_template.py in two different environments (passing --hidden_dim 500 from the command line in both cases).  The seed is set to the same value in both versions, so I would have expected things to match better

**Comment by mpariente** [comment]
> Just to be sure, @mpariente and @gwichern did you make the training deterministic before running this comparison?

No, my bad. But I did make each run twice, and have consistent difference between 0.6.0 and 0.7.1

**Comment by ethanwharris** [comment]
It looks like the MNIST example doesn't set shuffle=True in [the dataloader](https://github.com/PyTorchLightning/pytorch-lightning/blob/3c2fd560aa4d31b4f48ee225b83361deec53d9c7/pl_examples/basic_examples/lightning_module_template.py#L201), that could be the cause of the poor MNIST performance

**Comment by williamFalcon** [comment]
ok, yeah... finding the same. Let's dig into this a bit.
The problem is that the core logic didn't really change though.
Might be dataloader related

**Comment by williamFalcon** [comment]
@PyTorchLightning/core-contributors

**Comment by awaelchli** [comment]
the @pl.dataloader decorator got removed in 0.7. could it be related? (can't test right now).

**Comment by mpariente** [comment]
I didn't try to find the cause of this yet, I thought I should report first. 


> @mpariente is data being shuffled in your case?
Yes it is.

**Comment by williamFalcon** [comment]
might be the refresh rate of the progress bar. maybe that also changed the update freq of the loggers by mistake

**Comment by mpariente** [comment]
> might be the refresh rate of the progress bar. maybe that also changed the update freq of the loggers by mistake

Had considered this possibility as well. But for a given epoch (while training), the results are significantly degraded.

**Comment by williamFalcon** [comment]
Using 0.6.0
At the end of 1 epoch.

![image](https://user-images.githubusercontent.com/3640001/76650995-13a04300-653a-11ea-9cbb-4f7e2f5954b5.png)


Using 0.7.1

![image](https://user-images.githubusercontent.com/3640001/76651055-34689880-653a-11ea-8bfd-709c83eae3dc.png)



Removing the decorators has no effect

![image](https://user-images.githubusercontent.com/3640001/76651131-519d6700-653a-11ea-898a-29185bc7d7ec.png)

**Comment by williamFalcon** [comment]
And when doing exactly the same model including environment reset we get an exact curve match...

![image](https://user-images.githubusercontent.com/3640001/76651947-eeaccf80-653b-11ea-8c24-758cea9e51ba.png)

**Comment by mpariente** [comment]
We'll try to see if we get the same problem when we are setting seeds. 
Thanks for taking time for it.
If it persists, I'll try to set up a reproducible example but the dataset we use is not open, it makes things complicated.

**Comment by williamFalcon** [comment]
any news @mpariente

**Comment by mpariente** [comment]
Real training takes over 2days so I was finishing both trainings. 
Differences are still huge and it didn't compensate over time. Didn't try to set the seeds yet. I'll let you know, I don't have a lot of compute power available for these experiments.
![Screenshot from 2020-03-15 13-17-26](https://user-images.githubusercontent.com/18496796/76701212-60f9ed00-66bf-11ea-967e-c86e567d8828.png)

We consider it an issue on our end and downgraded to version 0.6.0 for now.

**Comment by williamFalcon** [comment]
@mpariente was thinking about this. maybe it has to do with distributedSampler if you're using that? since we now inject that automatically, it may be that your effective batch size is differnt now and thus if you use the same learning rate, you won't get the best results?

It might be that you have to readjust your learning rate. (from this graph, i would make it slower).

**Comment by mpariente** [comment]
Hmm, we use `dp` only, not `ddp`, would this still apply? 
Thanks for taking a look again

**Comment by dreamgonfly** [comment]
Seen the same performance drop moving from 0.6.0 to 0.7.1 on my application. This is the biggest obstacle preventing me from upgrading to newer version.

We use `dp` only.

**Comment by mpariente** [comment]
Well, this persists. 
I finally took time to reproduce the issue, on another architecture. 
Now, the training is deterministic, here are the tensorboard logs (grey pl0.7.1, orange pl0.6., nothing else changes between the envs) : 
![Screenshot from 2020-03-28 21-28-30](https://user-images.githubusercontent.com/18496796/77833120-35551900-713b-11ea-9dc8-da018a60aa15.png)

Training is not over but the differences are already not negligible. 
I also have a video about the 10 first epochs if you'd like. 

This script to reproduce are [here](https://github.com/mpariente/asteroid/tree/pldet/egs/whamr/TasNet) but the training dataset is under license..

**Comment by williamFalcon** [comment]
ok awesome. will look into this.
and this was introduced in 0.7.1?

**Comment by mpariente** [comment]
IIRC 0.7.0 was not backward compatible because of `pl.data_loader` so I couldn't test it. 

> It's weird because the tests do test a specific performance goal

I know you're doing the best you can about this, no worries. 

For now, both architecture involved LSTMs, did you change anything about BBTT? 
I'm going to try with a ConvNet, see if it changes as well.

**Comment by williamFalcon** [comment]
ummm... i don't think we did but that's good to know. Maybe it is RNN related. 
Why don't you do the sample in this colab so we can unify efforts here.

I want to create the following tests:
1. MNIST using MLPs.
2. cifar10 using CNNs.
3. an RNN sample, maybe sequence classification. whatever the equivalent simple test is for RNN.
4. that's probably it.

**Comment by mpariente** [comment]
Tried on two convolutional architectures and the training and validation curves are a perfect match. 
At first sight, it seems to be RNN related, which would be a good news.

**Comment by mpariente** [comment]
Any update on that please?

**Comment by williamFalcon** [comment]
will do an rnn test. however, we now have a parity test between pytorch pure and lightning with convnets in continuous integration. the test forces a match across trials to 5 decimal points. 

i’ll add an rnn test as well

**Comment by mpariente** [comment]
Sounds great ! 
Could you give me a link to the parity tests you mentioned please?

**Comment by williamFalcon** [comment]
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/benchmarks/test_trainer_parity.py

**Comment by williamFalcon** [comment]
this runs on every PR to make sure no PR breaks parity.
We do have a bit of difference in speed with pytorch, but looking into it. It looks related to logging, tqdm and tensorboard.

so, speedwise it's not a fair comparison because the pure pytorch version has no logging or any of that, whereas lightning does

**Comment by mpariente** [comment]
Oh I didn't see the PR, I thought you'd ping this Issue with it. 
About the RNN, did you decide the task? Do you want a fake dataset or something real? I can put together an averaging RNN example if needed.

**Comment by williamFalcon** [comment]
yeah, that would be super helpful! maybe the addition task is a good dataset to test?

**Comment by awaelchli** [comment]
I took the code from #1351 and ran it for 10 epochs and 3 runs on CPU first, then I noticed that there is a performance gap between 0.6.0 and 0.7.1 in the third decimal point.
https://colab.research.google.com/drive/1yek1fUkIEmJgt9iI7pnr3HWFNgrf14pr#scrollTo=wezjddkD68rz
Not sure if it helps.

**Comment by mpariente** [comment]
Thanks for looking into this, could you grant me access to the colab please? 

Did you also try on GPU?

**Comment by awaelchli** [comment]
Try again, I had sharing turned off.
No, Colab doesn't want to give me GPU for some reason, that's why I tried CPU.

**Comment by mpariente** [comment]
Ok, I can reproduce the same results as you and checked that the pytorch vanilla_loop also passes, and this is the case. 
I'm trying something else, I'll let you know if it gives anything

**Comment by mpariente** [comment]
@williamFalcon have you seen this? 
I think the parity test are not as good as they could be because if `configure_optimizers` does something under the hood, it won't have any impact as these are the optimizers used in the pure pytorch case as well.


> I took the code from #1351 and ran it for 10 epochs and 3 runs on CPU first (because @mpariente also has no GPU), then I noticed that there is a performance gap between 0.6.0 and 0.7.1 in the third decimal point.
> https://colab.research.google.com/drive/1yek1fUkIEmJgt9iI7pnr3HWFNgrf14pr
> Not sure if it helps.

And the results @awaelchli mentionned shows exactly this right? Parity test passes but the performances are different, how can we explain that?

**Comment by jeremyjordan** [comment]
we should also probably include `truncated_bptt` in the parity test

**Comment by mpariente** [comment]
Sorry to ping you @williamFalcon, but this is not resolved.

**Comment by williamFalcon** [comment]
we have parity tests now with an exact performance match...
can you provide a colab where you can reproduce this behavior?

**Comment by mpariente** [comment]
> we have parity tests now with an exact performance match...
> can you provide a colab where you can reproduce this behavior?

I would not call it exact performance match actually : performance are matching between `pytorch-lightning` and `torch`, true. But if you change the version of `pytorch-lightning`, the perfs are changing and they shouldn't because the seeds are exactly the same. 

So something is happening under the hood, see [those lines](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/benchmarks/test_trainer_parity.py#L90-L94) for example, in 0.6.0, the train dataloader is also changed by lightning right? 

I don't think these parity tests are as valuable as they should be.

**Comment by williamFalcon** [comment]
the performance comparison has to be against pure pytorch because that’s the bound for speed and accuracy. comparing across lightning versions makes no sense. 

again, can’t help without a real example that breaks on colab. every other time anyone has brought up a performance difference has ended up finding a bug in their code

**Comment by mpariente** [comment]
> the performance comparison has to be against pure pytorch because that’s the bound for speed and accuracy. comparing across lightning versions makes no sense.
> 

But I don't think it qualifies as pure pytorch, everything still comes from a lightning module. 

> again, can’t help without a real example that breaks on colab. every other time anyone has brought up a performance difference they’ve ended up finding a bug in their code.
> 
> happy to fix if something is broken, but we need tangible proof to find a possible problem.

I understand, I'll try to build an example that fails next week, thanks again

**Comment by williamFalcon** [comment]
it’s literally the same code. it’s like saying 2 = (2) are different haha. it’s written this way for convenience because the pytorch code is exactly the same...

**Comment by mpariente** [comment]
I've tried with 0.7.5 against 0.6.0 and got the same results on several of our architectures. We'll finally upgrade and get all the new features you integrated :grinning: 
Thanks again for looking into it, I'm closing this.
